components/verifier/verifier_test.go

```diff
@@ -0,0 +1,263 @@
+	require.Nil(t, err)
+	kOp, err := keyStore.NewKey(pass)
+	require.Nil(t, err)
+	err = keyStore.UnlockKey(kOp, pass)
+	require.Nil(t, err)
+	is, err := issuer.New(cfg, kOp, []merkletree.Entrier{}, storage, keyStore, idenPubOnChain)
+	require.Nil(t, err)
+	return is, storage, keyStore
+}
+
+func mockInitState(t *testing.T, idenPubOnChain *idenpubonchain.IdenPubOnChainMock, is *issuer.Issuer, genesisState *merkletree.Hash) (*types.Transaction, *merkletree.Hash) {
+	var ethTx types.Transaction
+	newState, _ := is.State()
+	sig, err := is.SignBinary(issuer.SigPrefixSetState, append(genesisState[:], newState[:]...))
+	require.Nil(t, err)
+	idenPubOnChain.On("InitState", is.ID(), genesisState, newState, []byte(nil), []byte(nil), sig).Return(&ethTx, nil).Once()
+	return &ethTx, newState
+}
+
+func mockSetState(t *testing.T, idenPubOnChain *idenpubonchain.IdenPubOnChainMock, is *issuer.Issuer, oldState *merkletree.Hash) (*types.Transaction, *merkletree.Hash) {
+	var ethTx types.Transaction
+	newState, _ := is.State()
+	sig, err := is.SignBinary(issuer.SigPrefixSetState, append(oldState[:], newState[:]...))
+	require.Nil(t, err)
+	idenPubOnChain.On("SetState", is.ID(), newState, []byte(nil), []byte(nil), sig).Return(&ethTx, nil).Once()
+	return &ethTx, newState
+}
+
+func newIssuerIssuedClaim(t *testing.T, idenPubOnChain *idenpubonchain.IdenPubOnChainMock, claim merkletree.Entrier) *issuer.Issuer {
+	is, _, _ := newIssuer(t, idenPubOnChain)
+	genesisState, _ := is.State()
+	err := is.IssueClaim(claim)
+	require.Nil(t, err)
+
+	_, newState := mockInitState(t, idenPubOnChain, is, genesisState)
+
+	// Publishing state for the first time
+	err = is.PublishState()
+	require.Nil(t, err)
+
+	blockN := uint64(12)
+	blockTs := int64(105000)
+	idenPubOnChain.On("GetState", is.ID()).Return(&proof.IdenStateData{IdenState: newState, BlockN: blockN, BlockTs: blockTs}, nil)
+	idenPubOnChain.On("GetStateByBlock", is.ID(), blockN).
+		Return(&proof.IdenStateData{BlockN: blockN, BlockTs: blockTs, IdenState: newState}, nil)
+
+	err = is.SyncIdenStatePublic()
+	require.Nil(t, err)
+
+	return is
+}
+
+func newIssuerIssuedClaim2(t *testing.T, idenPubOnChain *idenpubonchain.IdenPubOnChainMock, claim1, claim2 merkletree.Entrier) (*issuer.Issuer, *proof.CredentialExistence) {
+	is, _, _ := newIssuer(t, idenPubOnChain)
+	genesisState, _ := is.State()
+	err := is.IssueClaim(claim1)
+	require.Nil(t, err)
+
+	_, newState := mockInitState(t, idenPubOnChain, is, genesisState)
+
+	// Publishing state for the first time with claim1
+	err = is.PublishState()
+	require.Nil(t, err)
+
+	blockN := uint64(12)
+	blockTs := int64(100)
+	idenPubOnChain.On("GetState", is.ID()).Return(&proof.IdenStateData{IdenState: newState, BlockN: blockN, BlockTs: blockTs}, nil).Once()
+	idenPubOnChain.On("GetStateByBlock", is.ID(), blockN).
+		Return(&proof.IdenStateData{BlockN: blockN, BlockTs: blockTs, IdenState: newState}, nil)
+
+	err = is.SyncIdenStatePublic()
+	require.Nil(t, err)
```

Comment:
newIssuerIssuedClaim2 could call newIssuerIssuedClaim (or a new function that both use) in order to avoid repeating code.

Reply:
I will add a helper function to avoid repeating code in the next PR (I have already some commits in between).